scapy/contrib/socks.py

```diff
@@ -0,0 +1,168 @@
+# This file is part of Scapy
+# See http://www.secdev.org/projects/scapy for more informations
+# Copyright (C) Philippe Biondi <phil@secdev.org>
+# This program is published under a GPLv2 license
+
+# scapy.contrib.description = Socket Secure (SOCKS)
+# scapy.contrib.status = loads
+
+"""SOCKS4/5 Protocol"""
+
+from scapy.config import conf
+from scapy.error import warning
+from scapy.layers.dns import DNSStrField
+from scapy.layers.inet import TCP, UDP
+from scapy.layers.inet6 import IP6Field
+from scapy.fields import ByteField, ByteEnumField, ConditionalField,\
+    ShortField, IPField, StrField, MultipleTypeField
+from scapy.packet import Packet, bind_layers, bind_bottom_up
+
+# TODO: support the 3 different authentification exchange procedures for SOCKS5  # noqa: E501
+# 1 - Plain (https://tools.ietf.org/html/rfc1928 - 3.Procedure for TCP-based clients)  # noqa: E501
+# 2 - Username/password (https://tools.ietf.org/html/rfc1929)
+# 3 - GSS-API (https://tools.ietf.org/html/rfc1961)
+
+
+class SOCKS(Packet):
+    fields_desc = [
+        ByteEnumField("vn", 0x5,
+                      {0x4: "v4 - Request", 0x0: "v4 - Reply", 0x5: "v5"}),
+    ]
+
+    def guess_payload_class(self, pkt):
+        if self.underlayer and isinstance(self.underlayer, TCP):
+            if self.vn == 0x5:
+                if self.underlayer.dport == 1080:
+                    return SOCKS5Request
+                elif self.underlayer.sport == 1080:
+                    return SOCKS5Reply
+            elif self.vn == 0x4:
```

Comment:
I think we do not need to assume the port is 1080 here. Also, we might want a heuristic to tell which port is the server port. What do you think of this:
```Python
# define conf.socks_ports = [1080]

    def guess_payload_class(self, pkt):
        if self.vn == 0x4:
            return SOCKS4Request
        if self.vn == 0x0:
            return SOCKS4Reply
        if self.vn == 0x5 and isinstance(self.underlayer, TCP) and self.underlayer.sport in conf.socks_ports:
            return SOCKS5Reply
        return SOCKS5Request
```